modules/lang-painless/src/main/java/org/elasticsearch/painless/semantic/ResolveTypesPass.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.painless.semantic;
+
+import org.elasticsearch.painless.lookup.PainlessLookup;
+import org.elasticsearch.painless.node.ANode;
+import org.elasticsearch.painless.node.DResolvedType;
+import org.elasticsearch.painless.node.DUnresolvedType;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class ResolveTypesPass {
+
+    public interface Visitor {
+        void visit(ANode node, PainlessLookup lookup);
+    }
+
+    protected final PainlessLookup lookup;
+    protected final Map<Class<? extends ANode>, Visitor> visitors = new HashMap<>();
+
+    public ResolveTypesPass(PainlessLookup lookup) {
+        this.lookup = lookup;
+        buildVisitors();
+    }
+
+    private void buildVisitors() {
+        visitors.put(DUnresolvedType.class, (node, lookup) -> {
+            DUnresolvedType unresolvedType = (DUnresolvedType)node;
+            Class<?> type = lookup.canonicalTypeNameToType(unresolvedType.getUnresolvedType());
+
+            if (type == null) {
+                throw new IllegalArgumentException("cannot resolve symbol [" + unresolvedType.getUnresolvedType() + "]");
+            }
+
+            node.getParent().replace(unresolvedType, new DResolvedType(unresolvedType.getLocation(), type));
+        });
+    }
+
+    public void pass(ANode node) {
+        Visitor visitor = visitors.get(node.getClass());
```

Comment:
This looks like it will require every type of semantic pass we write to implement iteration over the tree. Could we instead have this built into the base class so that these define what types of nodes they want to operate on?